File: packages/numlang-rs/tests/unit.rs
use numlang::{abbreviate_unit, expand_unit, unit_type, UnitType};

#[test]
fn test_expand_unit_basic() {
    assert_eq!(expand_unit("ml"), Some("milliliter"));
    assert_eq!(expand_unit("MG"), Some("milligram"));
    assert_eq!(expand_unit("kg"), Some("kilogram"));
    assert_eq!(expand_unit("oz"), Some("ounce"));
    assert_eq!(expand_unit("IU"), Some("international unit"));
    assert_eq!(expand_unit("mcg"), Some("microgram"));
    assert_eq!(expand_unit("μg"), Some("microgram"));
    assert_eq!(expand_unit("cc"), Some("cubic centimeter"));
    assert_eq!(expand_unit("gtt"), Some("drop"));
    assert_eq!(expand_unit("tab"), Some("tablet"));
    assert_eq!(expand_unit("cap"), Some("capsule"));
    assert_eq!(expand_unit("patch"), Some("patch"));
    assert_eq!(expand_unit("dose"), Some("dose"));
    assert_eq!(expand_unit("unknown"), None);
}

#[test]
fn test_expand_unit_plural_and_case() {
    assert_eq!(expand_unit("tab"), Some("tablet"));
    assert_eq!(expand_unit("cap"), Some("capsule"));
    assert_eq!(expand_unit("chew"), Some("chew"));
}

#[test]
fn test_abbreviate_unit_singular() {
    assert_eq!(abbreviate_unit("milliliter"), Some("ml"));
    assert_eq!(abbreviate_unit("drop"), Some("gtt"));
    let abbr = abbreviate_unit("microgram");
    assert!(abbr == Some("mcg") || abbr == Some("μg"));
}

#[test]
fn test_abbreviate_unit_plural() {
    assert_eq!(abbreviate_unit("milliliters"), Some("ml"));
    assert_eq!(abbreviate_unit("drops"), Some("gtt"));
    let abbr = abbreviate_unit("micrograms");
    assert!(abbr == Some("mcg") || abbr == Some("μg"));
}

#[test]
fn test_abbreviate_unit_case_insensitive() {
    assert_eq!(abbreviate_unit("Milliliters"), Some("ml"));
}

#[test]
fn test_abbreviate_unit_unknown() {
    assert_eq!(abbreviate_unit("unknownunit"), None);
    assert_eq!(abbreviate_unit(""), None);
}

#[test]
fn test_unit_type_basic() {
    assert_eq!(unit_type("mg"), Some(UnitType::Mass));
    assert_eq!(unit_type("ml"), Some(UnitType::Volume));
    assert_eq!(unit_type("tablet"), Some(UnitType::DosageForm));
    assert_eq!(unit_type("patch"), Some(UnitType::DosageForm));
    assert_eq!(unit_type("oz"), Some(UnitType::Mass));
    assert_eq!(unit_type("cm"), Some(UnitType::Length));
    assert_eq!(unit_type("hr"), Some(UnitType::Time));
    assert_eq!(unit_type("unknown"), None);
}

#[test]
fn test_unit_type_expanded_and_plural() {
    assert_eq!(unit_type("milligram"), Some(UnitType::Mass));
    assert_eq!(unit_type("milliliters"), Some(UnitType::Volume));
    assert_eq!(unit_type("tablets"), Some(UnitType::DosageForm));
    assert_eq!(unit_type("patches"), Some(UnitType::DosageForm));
    assert_eq!(unit_type("ounces"), Some(UnitType::Mass));
    assert_eq!(unit_type("centimeters"), Some(UnitType::Length));
    assert_eq!(unit_type("hours"), Some(UnitType::Time));
    assert_eq!(unit_type("unknownunit"), None);
}

File: packages/numlang-js/src/unit.ts
const unitMap: Record<string, string> = {
  // Volume
  ml: "milliliter",
  cc: "cubic centimeter",
  l: "liter",
  dl: "deciliter",
  ul: "microliter",
  "fl oz": "fluid ounce",
  gal: "gallon",
  qt: "quart",
  pt: "pint",
  tbsp: "tablespoon",
  tsp: "teaspoon",
  gtt: "drop",
  spray: "spray",
  puff: "puff",
  "mg/kg": "milligram per kilogram",
  "ml/kg": "milliliter per kilogram",
  "mcg/kg": "microgram per kilogram",
  "ug/kg": "microgram per kilogram",

  // Mass/weight
  mg: "milligram",
  g: "gram",
  kg: "kilogram",
  mcg: "microgram",
  μg: "microgram",
  lb: "pound",
  oz: "ounce",
  iu: "international unit",
  u: "unit",
  meq: "milliequivalent",

  // Length
  mm: "millimeter",
  cm: "centimeter",
  m: "meter",
  km: "kilometer",
  in: "inch",
  ft: "foot",
  yd: "yard",
  mi: "mile",

  // Time
  s: "second",
  sec: "second",
  min: "minute",
  h: "hour",
  hr: "hour",
  d: "day",
  wk: "week",
  mo: "month",
  yr: "year",

  // Dosage forms
  tab: "tablet",
  cap: "capsule",
  supp: "suppository",
  amp: "ampule",
  vial: "vial",
  patch: "patch",
  chew: "chew",
  dose: "dose",
  sachet: "sachet",
  pump: "pump",
  app: "application",
  pipette: "pipette",
};

function toPlural(s: string): string {
  // Simple pluralisation for units
  if (s.endsWith("y")) return s.slice(0, -1) + "ies";
  if (
    s.endsWith("ch") ||
    s.endsWith("sh") ||
    s.endsWith("s") ||
    s.endsWith("x") ||
    s.endsWith("z")
  )
    return s + "es";
  return s + "s";
}

function reverseUnitMap(): Record<string, string> {
  const m: Record<string, string> = {};
  for (const [abbr, expanded] of Object.entries(unitMap)) {
    m[expanded] = abbr;
    m[toPlural(expanded)] = abbr;
  }
  return m;
}

const reverseMap = reverseUnitMap();

export function expandUnit(unit: string): string | undefined {
  const key = unit.trim().toLowerCase();
  return unitMap[key];
}

export function abbreviateUnit(expanded: string): string | undefined {
  const key = expanded.trim().toLowerCase();
  return reverseMap[key];
}
